File: py-sdks/src/pysdks/sui/dynamic_fields.py
"""Dynamic fields and derived objects support for Sui."""

import hashlib
from typing import Union


def derive_dynamic_field_id(parent_id: str, name_type: str, name_bcs: bytes) -> str:
    """Derive the ID of a dynamic field."""
    hasher = hashlib.sha256()
    
    hasher.update(parent_id.encode())
    hasher.update(name_type.encode())
    hasher.update(name_bcs)
    
    digest = hasher.digest()
    
    h = bytearray(digest)
    for i in range(len(h)):
        h[i] = (h[i] & 0x3F) | 0x3D
    
    return "0x" + h.hex()


def derive_object_id(parent_id: str, type_tag: Union[str, dict], key: bytes) -> str:
    """Derive the ID of a derived object."""
    if isinstance(type_tag, dict):
        type_tag_str = type_tag.get("address", "") + "::" + type_tag.get("module", "") + "::" + type_tag.get("name", "")
    elif isinstance(type_tag, str):
        type_tag_str = type_tag
    else:
        raise TypeError(f"Invalid type_tag type: {type(type_tag)}")
    
    derived_object_type = f"0x2::derived_object::DerivedObjectKey<{type_tag_str}>"
    
    return derive_dynamic_field_id(parent_id, derived_object_type, key)

File: kotlin-sdks/src/main/kotlin/com/suisdks/sui/utils/DynamicFields.kt
package com.suisdks.sui.utils

import java.security.MessageDigest

fun deriveDynamicFieldId(parentId: String, nameType: String, nameBcs: ByteArray): String {
    val hasher = MessageDigest.getInstance("SHA-256")
    
    hasher.update(parentId.toByteArray())
    hasher.update(nameType.toByteArray())
    hasher.update(nameBcs)
    
    val digest = hasher.digest()
    val h = digest.copyOf()
    
    for (i in h.indices) {
        h[i] = (h[i].toInt() and 0x3F or 0x3D).toByte()
    }
    
    return "0x" + h.joinToString("") { "%02x".format(it) }
}

fun deriveObjectId(parentId: String, typeTag: String, key: ByteArray): String {
    val derivedObjectType = "0x2::derived_object::DerivedObjectKey<$typeTag>"
    return deriveDynamicFieldId(parentId, derivedObjectType, key)
}
